src/main/java/net/neoforged/neoforge/common/IOUtilities.java

```diff
@@ -0,0 +1,160 @@
+            for (var file : filesToDelete.toList()) {
+                Files.deleteIfExists(file);
+            }
+        }
+    }
+
+    private static BiPredicate<Path, BasicFileAttributes> createPredicate(@Nullable String prefix) {
+        return (file, attributes) -> {
+            final var fileName = file.getFileName().toString();
+            return fileName.endsWith(TEMP_FILE_SUFFIX) && (prefix == null || fileName.startsWith(prefix));
+        };
+    }
+
+    /**
+     * Behaves much the same as {@link NbtIo#writeCompressed(CompoundTag, Path)},
+     * but uses {@link #atomicWrite(Path, WriteCallback)} behind the scenes to
+     * ensure the data is stored resilliently.
+     *
+     * @param tag  The tag to write.
+     * @param path The path to write the NBT to.
+     *
+     * @throws IOException if an I/O error occurs during writing.
+     */
+    public static void writeNbtCompressed(CompoundTag tag, Path path) throws IOException {
+        atomicWrite(path, stream -> {
+            try (var bufferedStream = new BufferedOutputStream(stream)) {
+                NbtIo.writeCompressed(tag, bufferedStream);
+            }
+        });
+    }
+
+    /**
+     * Behaves much the same as {@link NbtIo#write(CompoundTag, Path)},
+     * but uses {@link #atomicWrite(Path, WriteCallback)} behind the scenes to
+     * ensure the data is stored resilliently.
+     *
+     * @param tag  The tag to write.
+     * @param path The path to write the NBT to.
+     *
+     * @throws IOException if an I/O error occurs during writing.
+     */
+    public static void writeNbt(CompoundTag tag, Path path) throws IOException {
+        atomicWrite(path, stream -> {
+            try (var bufferedStream = new BufferedOutputStream(stream);
+                    var dataStream = new DataOutputStream(bufferedStream)) {
+                NbtIo.write(tag, dataStream);
+            }
+        });
+    }
+
+    /**
+     * Writes data to the given path "atomically", such that a crash will not
+     * leave the the file containing corrupted or otherwise half-written data.
+     * <p>
+     * This method operates by creating a temporary file, writing to that file,
+     * and then moving the temporary file to the correct location after flushing.
+     * If a crash occurs during this process, the temporary file will be
+     * abandoned.
+     * <p>
+     * Furthermore, the stream passed to {@code writeCallback} is not buffered,
+     * and it is the handler's responsibility to implement any buffering on top
+     * of this method.
+     *
+     * @param targetPath    The desired path to write to.
+     * @param writeCallback A callback which receives the opened stream to write
+     *                      data to.
+     *
+     * @throws IOException if an I/O error occurs during writing.
+     */
+    public static void atomicWrite(Path targetPath, WriteCallback writeCallback) throws IOException {
```

Comment:
An alternative pattern is to make a custom `OutputStream`. See https://github.com/neoforged/ModDevGradle/blob/912978861241d87483713f3ced664f6316f0d005/src/main/java/net/neoforged/moddevgradle/internal/utils/FileUtils.java#L42. As you prefer, I don't mind either solution.

Reply:
I'm not a huge fan of the custom OutputStream approach, since part of me worries about poorly-implemented inner/outer streams, as well as consumers which don't properly call close